server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java

```diff
@@ -0,0 +1,361 @@
+        return add(hash);
+    }
+
+    /**
+     * Add's the provided value to the set for tracking
+     */
+    public boolean add(byte[] value) {
+        MurmurHash3.Hash128 hash = MurmurHash3.hash128(value, 0, value.length, 0, new MurmurHash3.Hash128());
+        return add(hash);
+    }
+
+    /**
+     * Add's the provided value to the set for tracking
+     */
+    public boolean add(long value) {
+        return add(Numbers.longToBytes(value));
+    }
+
+    private boolean add(MurmurHash3.Hash128 hash) {
+        if (isSetMode) {
+            hashes.add(hash);
+            if (hashes.size() > threshold) {
+                convert();
+            }
+            return true;
+        }
+
+        boolean success = filters.get(filters.size() - 1).add(hash);
+        if (success == false) {
+            // filter is full, create a new one and insert there
+            CuckooFilter t = new CuckooFilter(capacity, fpp, rng);
+            t.add(hash);
+            filters.add(t);
+            breaker.accept(t.getSizeInBytes()); // make sure we account for the new filter
+        }
+        return true;
+    }
+
+    /**
+     * If we still holding values in a set, convert this filter into an approximate, cuckoo-backed filter.
+     * This will create a list of CuckooFilters, and null out the set of hashes
+     */
+    private void convert() {
+        if (isSetMode) {
+            long oldSize = getSizeInBytes();
+
+            filters = new ArrayList<>();
+            CuckooFilter t = new CuckooFilter(capacity, fpp, rng);
+            hashes.forEach(t::add);
+            filters.add(t);
+
+            hashes = null;
+            isSetMode = false;
+
+            breaker.accept(-oldSize); // this zeros out the overhead of the set
+            breaker.accept(getSizeInBytes()); // this adds back in the new overhead of the cuckoo filters
+        }
+    }
+
+    /**
+     * Get the approximate size of this datastructure.  Approximate because only the Set occupants
+     * are tracked, not the overhead of the Set itself.
+     */
+    public long getSizeInBytes() {
+        long bytes = 0;
+        if (hashes != null) {
+            bytes = (hashes.size() * 16) + 8 + 4 + 1;
+        }
+        if (filters != null) {
+            bytes += filters.stream().mapToLong(CuckooFilter::getSizeInBytes).sum();
+        }
+        return bytes;
+    }
+
+
+    /**
+     * Merge `other` cuckoo filter into this cuckoo.  After merging, this filter's state will
+     * be the union of the two.  During the merging process, the internal Set may be upgraded
+     * to a cuckoo if it goes over threshold
+     */
+    public void merge(SetBackedScalingCuckooFilter other) {
+        if (isSetMode && other.isSetMode) {
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            if (hashes.size() > threshold) {
+                convert();
```

Comment:
The `convert()` method assumes that `cuckoo_capacity > 2 * set_threshold` but this is not enforced anywhere in this class. I wonder if we should either enforce it (seems easiest solution to me) or make sure the `convert()` method can cope with the filter getting full during the conversion (i.e. create multiple filters during the loop)

Reply:
Yeah agreed, I can throw some exceptions and/or asserts to guarantee they are the same.  Was hardcoded so I just ignored it... but in case that changes in the future would be good to have checks here